File: trains-gdx-core/src/main/java/com/github/caay2000/trains/world/City.java
package com.github.caay2000.trains.world;

import com.github.caay2000.trains.common.Position;

import java.util.HashSet;
import java.util.Set;

public class City implements WorldObject {

    private final Position position;
    private final int population;
    private boolean connected;

    private Set<City> citiesInRange;

    public City(float x, float y, int population) {
        this.position = new Position(x, y);
        this.population = population;
        this.connected = false;
    }

    @Override
    public Position getPosition() {
        return position;
    }

    public int getPopulation() {
        return population;
    }

    public boolean isConnected() {
        return connected;
    }

    public boolean isNotConnected() {
        return !isConnected();
    }

    @Override
    public float distanceTo(WorldObject to) {
        return position.distanceTo(to.getPosition());
    }

    public Set<City> getCitiesInRange(Set<City> cities) {
        if (citiesInRange == null) {
            citiesInRange = new HashSet<>();
            for (City city : cities) {
                if (city.distanceTo(this) <= WorldConfiguration.maxRouteDistance) {
                    citiesInRange.add(city);
                }
            }
        }
        return citiesInRange;
    }

    public void connected() {
        this.connected = true;
    }
}

File: trains-gdx-core/src/main/kotlin/com/github/caay2000/trains/world/City.kt
package com.github.caay2000.trains.world

import com.github.caay2000.trains.Configuration
import com.github.caay2000.trains.debug

class City(val name: String, override val position: Position, population: Int) : Location {

    override val locationsInRange: Set<Location> = mutableSetOf()
    override val connections: Set<Location> = mutableSetOf()

    override val offer: Map<CargoType, Cargo> = hashMapOf()
    override val demand: Set<CargoType> = mutableSetOf(CargoType.PAX, CargoType.MAIL)

    var population: Int = population
        private set

    private var populationDelta: Float = (1..1000).random().toFloat().div(1000)
    private var paxDelta: Float = 0f

    override fun update(delta: Float) {

        increaseOffer(delta)
        increasePopulation(delta)
    }

    private fun increaseOffer(delta: Float) {

        if (connected()) {
            val cargo = (this.offer as MutableMap).getOrPut(CargoType.PAX) { Cargo(CargoType.PAX) }
            if (cargo.quantity < this.population * Configuration.cityMaxRatioPAXPopulation) {
                this.paxDelta += this.population * delta * Configuration.cityRatioPopulationPAX
                if (this.paxDelta.toInt() > 0) {
                    val grow = this.paxDelta.toInt()
                    cargo.add(grow)
                    this.paxDelta -= this.paxDelta.toInt()
                    // debug(grow > 0) { "$name pax available increased by $grow to $cargo" }
                }
            }
        }
    }

    fun deliverCargo(delivered: Int) {
        population += delivered
    }

    private fun increasePopulation(delta: Float) {
        populationDelta += delta * cityGrowingRatio()
        val grow = this.populationDelta.toInt()
        this.population += grow
        this.populationDelta -= this.populationDelta.toInt()

        // debug(grow > 0) { "$name populationIncreased by $grow to $population" }
    }

    private fun cityGrowingRatio() =
        if (connected()) Configuration.cityRatioGrowingPopulationConnected
        else Configuration.cityRatioGrowingPopulationNotConnected
}
